pkg/tsdb/cloudmonitoring/cloudmonitoring.go

```diff
@@ -66,145 +72,205 @@ const (
 	}
 
-	queryResult.Meta.Set("defaultProject", gceDefaultProject)
-	result.Results[refID] = queryResult
-
-	return result, nil
+	return &backend.QueryDataResponse{
+		Responses: backend.Responses{
+			req.Queries[0].RefID: {
+				Frames: data.Frames{data.NewFrame("").SetMeta(&data.FrameMeta{
+					Custom: map[string]interface{}{
+						"defaultProject": gceDefaultProject,
+					},
+				})},
+			},
+		},
+	}, nil
 }
 
-//nolint: staticcheck // plugins.DataPlugin deprecated
-func (e *Executor) executeTimeSeriesQuery(ctx context.Context, tsdbQuery plugins.DataQuery) (
-	plugins.DataResponse, error) {
-	result := plugins.DataResponse{
-		//nolint: staticcheck // plugins.DataPlugin deprecated
-		Results: make(map[string]plugins.DataQueryResult),
-	}
-
-	queryExecutors, err := e.buildQueryExecutors(tsdbQuery)
+func (s *Service) executeTimeSeriesQuery(ctx context.Context, req *backend.QueryDataRequest, dsInfo datasourceInfo) (
+	*backend.QueryDataResponse, error) {
+	resp := backend.NewQueryDataResponse()
+	queryExecutors, err := s.buildQueryExecutors(req)
 	if err != nil {
-		return plugins.DataResponse{}, err
+		return resp, err
 	}
 
 	for _, queryExecutor := range queryExecutors {
-		queryRes, resp, executedQueryString, err := queryExecutor.run(ctx, tsdbQuery, e)
+		queryRes, dr, executedQueryString, err := queryExecutor.run(ctx, req, s, dsInfo)
 		if err != nil {
-			return plugins.DataResponse{}, err
+			return resp, err
 		}
-		err = queryExecutor.parseResponse(&queryRes, resp, executedQueryString)
+		err = queryExecutor.parseResponse(queryRes, dr, executedQueryString)
 		if err != nil {
 			queryRes.Error = err
 		}
 
-		result.Results[queryExecutor.getRefID()] = queryRes
+		resp.Responses[queryExecutor.getRefID()] = *queryRes
 	}
 
-	return result, nil
+	return resp, nil
 }
 
-func (e *Executor) buildQueryExecutors(tsdbQuery plugins.DataQuery) ([]cloudMonitoringQueryExecutor, error) {
-	cloudMonitoringQueryExecutors := []cloudMonitoringQueryExecutor{}
-
-	startTime, err := tsdbQuery.TimeRange.ParseFrom()
+func queryModel(query backend.DataQuery) (grafanaQuery, error) {
+	var rawQuery map[string]interface{}
+	err := json.Unmarshal(query.JSON, &rawQuery)
 	if err != nil {
-		return nil, err
+		return grafanaQuery{}, err
 	}
 
-	endTime, err := tsdbQuery.TimeRange.ParseTo()
+	if rawQuery["metricQuery"] == nil {
+		// migrate legacy query
+		var mq metricQuery
+		err = json.Unmarshal(query.JSON, &mq)
+		if err != nil {
+			return grafanaQuery{}, err
+		}
+
+		return grafanaQuery{
+			QueryType:   metricQueryType,
+			MetricQuery: mq,
+		}, nil
+	}
+
+	var q grafanaQuery
+	err = json.Unmarshal(query.JSON, &q)
 	if err != nil {
-		return nil, err
+		return grafanaQuery{}, err
 	}
 
+	return q, nil
+}
+
+func (s *Service) buildQueryExecutors(req *backend.QueryDataRequest) ([]cloudMonitoringQueryExecutor, error) {
+	var cloudMonitoringQueryExecutors []cloudMonitoringQueryExecutor
+	startTime := req.Queries[0].TimeRange.From
```

Comment:
probably not important, I checked the size of query to avoid empty ones that finish by panic, but not sure about whether we actually control it already on front-end. #notrust

Reply:
I added a check to the beginning of QueryData so it should be caught by this point 🤞